specification/trace/semantic_conventions/http.md

```diff
@@ -194,6 +194,30 @@ Following attributes MUST be provided **at span creation time** (when provided a
 
 Note that in some cases host and port identifiers in the `Host` header might be different from the `server.address` and `server.port`, in this case instrumentation MAY populate `Host` header on `http.request.header.host` attribute even if it's not enabled by user.
 
+### HTTP client span duration
+
+HTTP client spans SHOULD start sometime before the first request byte is sent. This may or may not include connection time.
+
+HTTP client spans SHOULD end when the response body has been fully read or the response closes with an error.
+
+However, if there is any possibility for application code to not fully read the HTTP response
+(and for the HTTP client library to then have to clean up the HTTP response asynchronously),
+the HTTP client span MUST be ended earlier (e.g. after the HTTP response headers are read, or fail to be read).
+
+This avoids having to end the span asynchronously later on at a time which is no longer directly associated with the application code which made the HTTP request.
+
+Because of the potential for confusion around this, HTTP client library instrumentations SHOULD document their behavior around ending HTTP client spans.
+
+> **Note**
+> It is nice for HTTP client spans to represent the complete lifetime of the HTTP request.
+> An HTTP request ends once the response body has been fully read or the response closes with an error.
+> However, if there is any possibility for application code to not fully read the HTTP response
+> (and for the HTTP client library to then have to clean up the HTTP response asynchronously),
+> the HTTP client span SHOULD be ended earlier (e.g. after the HTTP response headers are read, or fail to be read).
+> This avoids having to end the span asynchronously later on at a time which is no longer directly associated with the application code which made the HTTP request.
+>
+> Because of the potential for confusion around this, HTTP client library instrumentations SHOULD document their behavior around ending HTTP client spans.
```

Comment:
I still think that defining separate and unambiguous metrics (and span kinds?) for different concepts would be better than having a single metric where the definition depends on API shape and has to be documented by individual libraries.

ie. if we would define separate:
- `http.client.latency` + "Http client latency spans"
- `http.client.duration` + "Http client duration spans"

and let libraries choose which of these two they implement (`latency`, `duration` or both), then the instrument name alone would make the definition clear without the need to dig into library-specific docs.

If this is impossible or impractical I'm keen to know why.

Reply:
> I think having a single metric that matches the HTTP client span duration best fits the cross-signal nature of semantic conventions.

+1 👍

Reply:
> a single metric that matches the HTTP client span duration

> a metric that consistently corresponds to the span duration 

But why isn't it possible to define two kinds of span durations?

Reply:
There can only be one span duration, but one could add span events to mark certain points in time.

Reply:
More precisely: I meant two kinds of Http client spans. But probably miss a lot of context on how the standard is being architectured.

Reply:
You mean that multiple nested spans would be created to delineate the different timings? That would in principle be possible from the OTel side, but seems overengineered and also very hard to impossible to implement when it comes to the actual library instrumentations.

Reply:
@antonfirsov, @trask had a similar idea that the evolution of the current state of things that we can add attributes?/events? on duration span for latency and this would allow us to derive both metrics if necessary.
Assuming the current `http.client.duration` means the longest we can reliably measure, means we can stamp more times on it without changing its meaning. 
(I hope I captured your idea correctly, @trask)

So, we're totally considering different options, but defining the right semantics needs time, effort, prototyping, community feedback, etc. So the proposal is to clarify the `http.client.duration` to allow additive changes in future to add more stuff. We're in full agreement that the current thing should be improved.

Reply:
> So the proposal here is to clarify the `http.client.duration` now to allow additive changes in future to add more stuff. We're in full agreement that the current thing should be improved.

👍 The current goal is to lay a stable foundation, and then we can then start layering in lower-level network details.

> the right semantics needs time, effort, prototyping, community feedback, etc

e.g. Java Netty HTTP client instrumentation currently supports experimental opt-in spans to capture some lower-level networking details (in addition to the HTTP client span)
* DNS resolution
* Connection
* SSL handshake

Reply:
> Assuming the current http.client.duration means the longest we can reliably measure

OK, that would be the logical consequence of my proposal of events, but thinking about it like that, this doesn't feel right. Then we would have to have the time until last byte, which in most cases would be misleading (even if we can reliably measure it.
Which means that for technial reaons, events would not be possible in all cases, which means nested spans would be a sensible option after all.

Reply:
@Oberon00 
with this PR, generic instrumentations SHOULD NOT measure time-to-last-byte if

> If there is any possibility for application code to not fully read the HTTP response
(and for the HTTP client library to then have to clean up the HTTP response asynchronously),
the HTTP client span SHOULD NOT be ended in this cleanup phase

essentially only manual or higher-level (client library) instrumentations should. I'd expect a  client library method that reads response fully before returning it to user to measure ttlb (with HTTP semantics or a custom one).

I also assume that in most cases, response bodies are rather small, and the difference between time-to-headers and ttlb is not important enough. I.e. extra nested spans might be justified.
Another option brought by Trask is attributes with multiple durations. Essentially we don't need extra spans as start time and attributes are the same.
Of course, this needs more thought and feedback, but it seems we have at least a couple of paths forward.

Do you have any other options allowing us to express different timings? or how would you approach solving it?